portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -24,9 +34,74 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private ArrayList<String> messages;
```

Comment:
Try to initialize the variable with type instead of implementation.  

List is a type (interface),  and ArrayList is an implementation of List. That could be other implementations such as LinkedList. Let's say later on, you want to switch to LinkedList, that is gonna be hard as variable 'messages' is tightly coupled with ArrayList.